cookbooks/bcpc-hadoop/libraries/dir_creation.rb

```diff
@@ -12,34 +12,47 @@ def dir_creation(mode, dirs, home, perms)
 
       # path to creation script and caches
       jruby_script = File.join(file_cache, 'create_dirs.rb')
-      dir_cache = File.join(file_cache, "#{mode}_list")
-      quota_cache = File.join(file_cache, "#{mode}_quotas.yml")
+      dirinfo_cache = File.join(file_cache, "#{mode}_dirinfo.yml")
 
-      # save dir cache
-      File.open(dir_cache, 'w+') { |f| f.puts(dirs) }
-      puts "dir cache(#{dirs.length}): #{dirs}"
+      # dir creation configuration attributes
+      dirinfo = node['bcpc']['hadoop']['hdfs'][mode]['dirinfo']
+      home = node['bcpc']['hadoop']['hdfs'][mode]['home']
+      puts "\n#{mode}_dir_creation(#{dirinfo.size}): #{dirinfo.keys.to_a}"
 
-      # build quota cache
-      quotas = dirs.map do |dir|
-        quota =
-          # use default quota if dir-specific quota unset
-          node['bcpc']['hadoop']['hdfs'][mode]["#{dir}_space_quota"] ||
-          node['bcpc']['hadoop']['hdfs'][mode]['space_quota']
-        [dir, quota]
+      # Build dirinfo cache
+      # -------------------
+      # use default space/ns quota if dir-specific quota unset
+      dirinfo_hash = dirinfo.map do |dir, info|
+        [
+          dir,
+          {
+            owner: info['owner'] ||
+              node['bcpc']['hadoop']['hdfs'][mode]['owner'],
+            group: info['group'] ||
+              node['bcpc']['hadoop']['hdfs'][mode]['group'],
+            perms: info['perms'] ||
+              node['bcpc']['hadoop']['hdfs'][mode]['perms'],
+            space_quota: info['space_quota'] ||
+              node['bcpc']['hadoop']['hdfs']["#{dir}_space_quota"] ||
+              node['bcpc']['hadoop']['hdfs'][mode]['space_quota'],
+            ns_quota: info['ns_quota'] ||
+              node['bcpc']['hadoop']['hdfs']["#{dir}_ns_quota"] ||
+              node['bcpc']['hadoop']['hdfs'][mode]['ns_quota']
+          }
+        ]
       end.to_h
```

Comment:
Move the hash merge out of the library and into recipe(s) to make rspec/chefspec work as expected when verifying the hash.